src/librustc_mir/interpret/intern.rs

```diff
@@ -0,0 +1,219 @@
+{
+    type V = MPlaceTy<'tcx>;
+
+    #[inline(always)]
+    fn ecx(&self) -> &CompileTimeEvalContext<'a, 'mir, 'tcx> {
+        &self.ecx
+    }
+
+    fn visit_primitive(&mut self, mplace: MPlaceTy<'tcx>) -> EvalResult<'tcx> {
+        // Go over all the primitive types
+        let ty = mplace.layout.ty;
+        if let ty::Ref(..) = ty.sty {
+            let value = self.ecx.read_immediate(mplace.into())?;
+            // Handle trait objects
+            if let Ok(meta) = value.to_meta() {
+                let layout = self.ecx.layout_of(ty.builtin_deref(true).unwrap().ty)?;
+                if layout.is_unsized() {
+                    if let ty::Dynamic(..) = self.ecx.tcx.struct_tail(layout.ty).sty {
+                        if let Ok(vtable) = meta.unwrap().to_ptr() {
+                            self.intern(vtable)?;
+                        }
+                    }
+                }
+            }
+            let mplace = self.ecx.ref_to_mplace(value)?;
+            // Check if we have encountered this pointer+layout combination before.
+            // Only recurse for allocation-backed pointers.
+            if mplace.ptr.is_ptr() {
+                match self.intern(mplace.ptr.to_ptr()?)? {
+                    // No need to recurse, these are interned already and statics may have
+                    // cycles, so we don't want to recurse there
+                    Some(IsStaticOrFn) => {},
+                    // intern everything referenced by this value
+                    None => self.ref_tracking.track(mplace, || ()),
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+pub fn intern_const_alloc_recursive(
+    ecx: &mut CompileTimeEvalContext<'a, 'mir, 'tcx>,
+    def_id: DefId,
+    ret: MPlaceTy<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+) -> EvalResult<'tcx> {
+    let tcx = ecx.tcx;
+    let frozen = ret.layout.ty.is_freeze(tcx.tcx, param_env, tcx.span);
+    let intern_mode = match tcx.is_static(def_id) {
+        Some(hir::Mutability::MutMutable) => InternMode::StaticMut,
+        Some(hir::Mutability::MutImmutable) => InternMode::Static { frozen },
+        None => InternMode::Const { frozen },
+    };
+
+    // type based interning
+
+    let mut ref_tracking = RefTracking::new(ret);
+    let mut leftover_allocations = FxHashSet::default();
+
+    // start with the outer most allocation
+    InternVisitor {
+        ref_tracking: &mut ref_tracking,
+        ecx,
+        mode: intern_mode,
+        leftover_relocations: &mut leftover_allocations,
+    }.intern(ret.ptr.to_ptr()?)?;
+
+    while let Some((mplace, ())) = ref_tracking.todo.pop() {
+        let interned = InternVisitor {
+            ref_tracking: &mut ref_tracking,
+            ecx,
+            mode: intern_mode,
+            leftover_relocations: &mut leftover_allocations,
+        }.visit_value(mplace);
+        if let Err(error) = interned {
+            if let EvalErrorKind::ValidationFailure(_) = error.kind {
```

Comment:
Where can this error originate from? The visitor itself does not throw a `ValidationFailure`.

I guess it's invalid enum discriminants? Please add a comment. Also, that may warrant a more specialized error message -- and a less cautious one. Seems like the errors we encounter here are not really unclear.